File: 133-clone-graph/leochang/133.cpp
/**
 * Definition for undirected graph.
 * struct UndirectedGraphNode {
 *     int label;
 *     vector<UndirectedGraphNode *> neighbors;
 *     UndirectedGraphNode(int x) : label(x) {};
 * };
 */
class Solution {
public:
	UndirectedGraphNode *cloneGraph(UndirectedGraphNode *node) {
		if (node == NULL) return node;

		if (hashMap.find(node) != hashMap.end()) {
			return hashMap[node];
		}

		UndirectedGraphNode *newNode = new UndirectedGraphNode(node->label);
		hashMap.insert(pair<UndirectedGraphNode*, UndirectedGraphNode*>(node, newNode));
		for (UndirectedGraphNode* neigh : node->neighbors) {
			hashMap[node]->neighbors.push_back(cloneGraph(neigh));
		}

		return hashMap[node];
	}
private:
	unordered_map<UndirectedGraphNode*, UndirectedGraphNode*> hashMap;
};

File: 133-clone-graph/ken/133.py
# Definition for a undirected graph node
# class UndirectedGraphNode:
#     def __init__(self, x):
#         self.label = x
#         self.neighbors = []

class Solution:
    # @param node, a undirected graph node
    # @return a undirected graph node
    def cloneGraph(self, node):
        if node is None:
            return None

        m = {}
        q = []
        q.append(node)
        newRoot = UndirectedGraphNode(node.label)
        m[node.label] = newRoot

        while len(q) > 0:
            srcNode = q.pop()
            dstNode = None

            if srcNode.label not in m:
                m[srcNode.label] = UndirectedGraphNode(srcNode.label)
            dstNode = m[srcNode.label]

            for neighbor in srcNode.neighbors:
                if neighbor.label not in m:
                    m[neighbor.label] = UndirectedGraphNode(neighbor.label)
                    q.append(neighbor)

                dstNode.neighbors.append(m[neighbor.label])

        return newRoot
